SConstruct

```diff
@@ -11,7 +11,7 @@ import SCons
 import SCons.Conftest as tests
 from SCons.Script.SConscript import SConsEnvironment
 
-DEFAULT_OPTIMISATION='s'  # compile with -Os
+DEFAULT_OPTIMISATION='2'  # compile with -O2
```

Comment:
I agree that this program is likely I/O-bound, and that it makes sense to have chosen the kind of optimization that results in the most noticeable improvement (in this case, *size* optimization rather than *speed* optimization). However, as you know and as you've implied, storage (or bandwidth) is considered very cheap these days, so maybe optimizing for size isn't an overall benefit to the user.

I wish that software were always optimized for speed instead, no matter how small the improvement. This is because I'm never running just one program; I'm never running *just* `rmlint`, and it seems to me that even tiny inefficiencies *accumulate* to produce noticeable degredation&mdash;no *one* program in particular is to blame, but the *group* as a whole *is* to blame.

<details>
<summary>These <code>bash</code> functions provide basic benchmarking.</summary>

    #!/bin/bash
    
    ITERATIONS=10
    
    msg() { printf "$@" >&3; }
    
    clean_up()
    {
        find /tmp -type d -not -path /tmp -prune -o \
            -type f \( -name out\*.json -o -name replay\*.json \) \
                -exec rm {} +
    
        find ./ -type d -not -path ./ -prune -o \
            -type f -name rmlint.\* \
                -exec rm {} +
    
        rm -f .rmlint.json
    }
    
    run_tests()
    {
        RM_TS_PRINT_CMD=1 RM_TS_PEDANTIC=0 nosetests -s -v -a '!slow'
    }
    
    warmup()
    {
        local compile="scons${1:+" $1"}"
        msg '%s\n' "$compile"
        msg 'Warmup... '
        git clean -fdx
        eval "$compile"
        run_tests
        clean_up
        msg 'finished\n'
    }
    
    benchmark()
    {
        msg 'Testing... 0'
        {
            local i
            time for ((i=1; i<=ITERATIONS; ++i)); do
                run_tests
                msg '\b%s' "$i"
            done 1>/dev/null 2>&1
        } 2>&1
        clean_up >/dev/null
        msg '\b\bfinished\n'
    }
    
    average_time()
    {
        {
            printf '%s\n' 'scale=3; total=0'
            while read clock time; do
                if [[ $clock =~ user|sys ]]; then
                    if [[ $time =~ ([^m]+)m([^s]+)s ]]; then
                        printf '%s\n' "total += ${BASH_REMATCH[1]}*60 + ${BASH_REMATCH[2]}"
                    else
                        return 1
                    fi
                fi
            done
            printf '%s\n' "total / $ITERATIONS"
        } | bc
    }
    
    test()
    (
        set -e
        exec 5>"$1" 4<"$1" 3>&1 2>/dev/null 1>&2
        shift
    
        warmup "$1"
        benchmark >&5
    
        msg '%s: %s seconds\n' 'Average time' "$(average_time <&4)"
    )
    
    separate()
    {
        (($#)) &&
            msg '\n%s\n' "=========== $1:" ||
            msg '%s\n' -----------
    }
    
    compare()
    (
        set -e
        exec 3>&1 2>/dev/null 1>&2
    
        if (($# > 2)); then
            msg '%s\n' 'Usage: compare [args1 [args2]]\n'
            return 1
        fi
    
        name1=${1%%:*}
        if [[ -z $name1 ]]; then
            name1=Test\ 1
        fi
        args1=${1#*:}
    
        name2=${2%%:*}
        if [[ -z $name2 ]]; then
            name2=Test\ 2
        fi
        args2=${2#*:}
    
        separate 'Setup'
        msg "Checking out \"$head\"... "
        git clean -fdx
        git checkout --detach "$head"
        msg 'finished\n'
    
        separate 'Running Tests'
        test /tmp/test1 "$args1" >&3
        if (($# != 2)); then
            separate "$name1"
            cat /tmp/test1 >&3
        else
            separate
            test /tmp/test2 "$args2" >&3
    
            separate "$name1"
            cat /tmp/test1 >&3
            separate "$name2"
            cat /tmp/test2 >&3
        fi
    )
</details><br>

The following commands build `rmlint` with the provided build-time arguments, and then calculate the average running time of the test suite across 10 iterations for that particular build:

    $ head=bff796458872a1781199cd72eaef30f7f891218d
    $ compare Size:O=s Speed:O=2
    
    =========== Setup:
    Checking out "bff796458872a1781199cd72eaef30f7f891218d"... finished
    
    =========== Running Tests:
    scons O=s
    Warmup... finished
    Testing... finished
    Average time: 77.231 seconds
    -----------
    scons O=2
    Warmup... finished
    Testing... finished
    Average time: 76.305 seconds
    
    =========== Size:
    
    real    28m14.593s
    user    9m33.759s
    sys     3m18.552s
    
    =========== Speed:
    
    real    27m53.960s
    user    9m25.998s
    sys     3m17.052s

So, when optimized for speed, the test suite is 0.926 seconds faster. That's not much for a human, but for a computer, that's a damn long time! I've got another little patch series in the works, which involves some inlining optimization that resulted from refactoring, and it might shave off a little more time:

    $ head=branch-name
    $ compare Speed:O=2
    
    =========== Setup:
    Checking out "branch-name"... finished
    
    =========== Running Tests:
    scons O=2
    Warmup... finished
    Testing... finished
    Average time: 75.920 seconds
    
    =========== Speed:
    
    real    27m50.822s
    user    9m25.630s
    sys     3m13.571s

That's a difference of 0.385 seconds.  
I wonder how many other fractions of a second could be shaved away.

Reply:
> I wish that software were always optimized for speed instead, no matter how small the improvement. This is because I'm never running just one program; I'm never running just rmlint, and it seems to me that even tiny inefficiencies accumulate to produce noticeable degredation—no one program in particular is to blame, but the group as a whole is to blame.

True, but that's only one metric to optimize. Each optimization is a trade-off. Sometimes you just trade CPU against memory. Sometimes you treat maintainability or being able to easily debug something against performance. Not claiming this is the case here, just saying that general goal *"more speed!"* shouldn't be always followed under all circumstances.

> These bash functions provide basic benchmarking.

Thank you for doing those. 

> The following commands build rmlint with the provided build-time arguments, and then calculate the average running time of the test suite across 10 iterations for that particular build:

Not sure if the test-suite is a very good benchmark for it. There's a questionable amount of time that goes into Python and each ``rmlint`` call usually only gets called on a tiny amount of data.
For quick and dirty benchmark I usually just copy some bigger file tree (parts of ``/usr``) on tmpfs and let it run there. What you want to test are longer running ``rmlint`` processes.

> That's a difference of 0.385 seconds.
> I wonder how many other fractions of a second could be shaved away.

Again, it's mostly I/O-bound and I'm not sure if your script really benchmarks the important bits of ``rmlint`` (i.e. the incremental hashing). Due to the high amount of calls to ``rmlint`` in the test suite the result might just be due to the faster startup (maybe faster command line processing and formatting). Not saying that the results themselves are not valid, but just saying that this might be a flaw in the benchmark.

But apart from this nitpicking: I guess you convinced me enough already to switch to ``-O2``. :smirk:

Reply:
Good points.

I'll note that *currently*, normal usage of `rmlint` might also involve a lot of startup costs; for instance, `rmlint --dedupe` may be called very many times indeed.